pkg/config/remote/service/subscriber_stream.go

```diff
@@ -0,0 +1,115 @@
+package service
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+
+	"github.com/DataDog/datadog-agent/pkg/config/remote/service/tuf"
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type subscriberStream struct {
+	mx                            sync.Mutex
+	rwmx                          sync.RWMutex
+	stream                        pbgo.AgentSecure_GetConfigUpdatesClient
+	agentClient                   pbgo.AgentSecureClient
+	currentConfigSnapshotVersions map[pbgo.Product]uint64
+	tufClient                     *tuf.DirectorPartialClient
+}
+
+func newSubscriberStream(streamCtx context.Context, conn *grpc.ClientConn) (*subscriberStream, error) {
+	c := &subscriberStream{
+		agentClient:                   pbgo.NewAgentSecureClient(conn),
+		tufClient:                     tuf.NewDirectorPartialClient(),
+		currentConfigSnapshotVersions: make(map[pbgo.Product]uint64),
+	}
+	c.startStream(streamCtx)
+	return c, nil
+}
+
+func (c *subscriberStream) startStream(streamCtx context.Context) {
+	var stream pbgo.AgentSecure_GetConfigUpdatesClient
+	var err error
+	for {
+		stream, err = c.agentClient.GetConfigUpdates(streamCtx)
+		if err != nil {
+			log.Errorf("Failed to establish channel to core-agent, retrying in %s...", errorRetryInterval)
+			time.Sleep(errorRetryInterval)
+			continue
+		} else {
+			log.Debugf("Successfully established channel to core-agent")
+			break
+		}
+	}
+	c.mx.Lock()
+	defer c.mx.Unlock()
+	c.stream = stream
```

Comment:
We should only reach this code path when there is no stream or a stream that was closed because of an error. And in that case I think they are GC'ed alright. I did a test where I ran this with `errorRetryInterval = 1ms` and took `-inuse_object` profiles once every second for a minute. I didn't see any memory leaking here.